src/components/CustomCheckout.tsx

```diff
@@ -0,0 +1,275 @@
+  ) => {
+    setContext((ctx) => {
+      if (ctx.stripe && ctx.customCheckoutSdk) {
+        return ctx;
+      }
+
+      return {stripe, customCheckoutSdk};
+    });
+  };
+
+  React.useEffect(() => {
+    let isMounted = true;
+
+    if (parsed.tag === 'async' && !ctx.stripe) {
+      parsed.stripePromise.then((stripe) => {
+        if (stripe && isMounted && !initCustomCheckoutCalled) {
+          // Only update context if the component is still mounted
+          // and stripe is not null. We allow stripe to be null to make
+          // handling SSR easier.
+          setInitCustomCheckoutCalled(true);
+          stripe.initCustomCheckout(options).then((customCheckoutSdk) => {
+            if (customCheckoutSdk) {
+              safeSetContext(stripe, customCheckoutSdk);
+              customCheckoutSdk.on('change', setSession);
+            }
+          });
+        }
+      });
+    } else if (
+      parsed.tag === 'sync' &&
+      parsed.stripe &&
+      !initCustomCheckoutCalled
+    ) {
+      setInitCustomCheckoutCalled(true);
+      parsed.stripe.initCustomCheckout(options).then((customCheckoutSdk) => {
+        if (customCheckoutSdk) {
+          safeSetContext(parsed.stripe, customCheckoutSdk);
+          customCheckoutSdk.on('change', setSession);
+        }
+      });
+    }
+
+    return () => {
+      isMounted = false;
+    };
+  }, [
+    parsed,
+    ctx,
+    options,
+    initCustomCheckoutCalled,
+    setInitCustomCheckoutCalled,
+    setSession,
+  ]);
+
+  // Warn on changes to stripe prop
+  const prevStripe = usePrevious(rawStripeProp);
+  React.useEffect(() => {
+    if (prevStripe !== null && prevStripe !== rawStripeProp) {
+      console.warn(
+        'Unsupported prop change on CustomCheckoutProvider: You cannot change the `stripe` prop after setting it.'
+      );
+    }
+  }, [prevStripe, rawStripeProp]);
+
+  // Apply updates to elements when options prop has relevant changes
+  const prevOptions = usePrevious(options);
+  React.useEffect(() => {
+    if (!ctx.customCheckoutSdk) {
+      return;
+    }
+
+    if (
+      options.clientSecret &&
+      !isUnknownObject(prevOptions) &&
+      !isEqual(options.clientSecret, prevOptions.clientSecret)
+    ) {
+      console.warn(
+        'Unsupported prop change: options.client_secret is not a mutable property.'
+      );
+    }
+
+    const previousAppearance = prevOptions?.elementsOptions?.appearance;
+    const currentAppearance = options?.elementsOptions?.appearance;
+    if (currentAppearance && !isEqual(currentAppearance, previousAppearance)) {
+      ctx.customCheckoutSdk.changeAppearance(currentAppearance);
+    }
+  }, [options, prevOptions, ctx.customCheckoutSdk]);
+
+  // Attach react-stripe-js version to stripe.js instance
+  React.useEffect(() => {
+    registerWithStripeJs(ctx.stripe);
+  }, [ctx.stripe]);
+
+  const customCheckoutContextValue = extractCustomCheckoutContextValue(
+    ctx.customCheckoutSdk
+  );
```

Comment:
This should be memoized. As is, each re-render of this component is deriving a new context value, causing all consumers to re-render.

Reply:
🤔  I think we actually want to recompute this value for ever re-render, so it can give the latest `ctx.customCheckoutSdk.session()` to its consumers.

We trigger the state change on 
https://github.com/stripe/react-stripe-js/blob/26bb4e222db70e47026f2fb03f99c72888374013/src/components/CustomCheckout.tsx#L133

https://github.com/stripe/react-stripe-js/blob/26bb4e222db70e47026f2fb03f99c72888374013/src/components/CustomCheckout.tsx#L147

Reply:
After discussing it with @danwang-stripe , I pushed a new commit to memorize the value https://github.com/stripe/react-stripe-js/pull/439/commits/5d96e44fc00e9aae51c12e9e6957be6dd6cbeb71